fix: File Upload widget disabled prop treats `undefined` as true

christian-byrne opened the pull request:
## Summary

Fixed `WidgetFileUpload` component disabled prop handling by ensuring `readonly || false` evaluation for button states.

## Changes

- **What**: Updated [WidgetFileUpload.vue](https://vuejs.org/guide/components/props.html#boolean-casting) disabled prop binding from `readonly` to `readonly || false`

## Review Focus

Boolean prop casting behavior in Vue components and disabled state consistency across all interactive elements.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5528-fix-File-Upload-widget-disabled-prop-treats-undefined-as-true-26d6d73d3650810ca1c9eb89f89d2f29) by [Unito](https://www.unito.io)

christian-byrne committed:
fix file upload widget disabled prop

DrJKL commented on src/renderer/extensions/vueNodes/widgets/components/WidgetFileUpload.test.ts:
```diff
@@ -0,0 +1,428 @@
+import WidgetFileUpload from './WidgetFileUpload.vue'
+
+describe('WidgetFileUpload File Handling', () => {
+  const createMockWidget = (
+    value: File[] | null = null,
+    options: Record<string, any> = {},
+    callback?: (value: File[] | null) => void
+  ): SimplifiedWidget<File[] | null> => ({
+    name: 'test_file_upload',
+    type: 'file',
+    value,
+    options,
+    callback
+  })
```
This is replicated in enough tests that it's probably worth extracting to a test utility, what do you think?

christian-byrne replied:
Done in https://github.com/Comfy-Org/ComfyUI_frontend/pull/5528/commits/f9d5f31897d2e6292afdef6058861e6a0ab64503

DrJKL commented on src/renderer/extensions/vueNodes/widgets/components/WidgetFileUpload.test.ts:
```diff
@@ -0,0 +1,428 @@
+      const wrapper = mountComponent(widget, null)
+
+      const fileInput = wrapper.find('input[type="file"]')
+      const clickSpy = vi.spyOn(fileInput.element as HTMLInputElement, 'click')
```
Type narrowing instead of type assertion?

christian-byrne replied:
Done in https://github.com/Comfy-Org/ComfyUI_frontend/pull/5528/commits/f9d5f31897d2e6292afdef6058861e6a0ab64503

DrJKL commented on src/renderer/extensions/vueNodes/widgets/components/WidgetFileUpload.test.ts:
```diff
@@ -0,0 +1,428 @@
+          // Should show normal upload UI when no specific preview is available
+          expect(wrapper.find('img').exists()).toBe(false)
+          expect(wrapper.find('.pi-volume-up').exists()).toBe(false)
+        }
```
I feel like conditionals in tests are an anti-pattern.
Could you split this by the expected into separate [it.for](https://vitest.dev/api/#test-for) scenarios?

christian-byrne replied:
Done in https://github.com/Comfy-Org/ComfyUI_frontend/pull/5528/commits/f9d5f31897d2e6292afdef6058861e6a0ab64503

DrJKL commented on src/renderer/extensions/vueNodes/widgets/components/WidgetFileUpload.test.ts:
```diff
@@ -0,0 +1,428 @@
+      const clickSpy = vi.spyOn(fileInput.element as HTMLInputElement, 'click')
+
+      // Find PrimeVue Button component with folder icon
+      const folderButton = wrapper.findComponent(Button)
```
If you need the component to exist, [getComponent](https://test-utils.vuejs.org/api/#getComponent) will throw the error for you

christian-byrne replied:
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5528/commits/5de42af20d5a427815518662a8813410f60df34b

DrJKL commented on src/renderer/extensions/vueNodes/widgets/components/WidgetFileUpload.test.ts:
```diff
@@ -0,0 +1,428 @@
+
+      // Should not throw error when method exists
+      const vm = wrapper.vm as any
+      if (vm.triggerFileInput) {
+        expect(() => vm.triggerFileInput()).not.toThrow()
+      }
```
Does this work?

```suggestion
      expect(() => vm.triggerFileInput?.()).not.toThrow()
```

christian-byrne replied:
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5528/commits/5de42af20d5a427815518662a8813410f60df34b

DrJKL reviewed (commented):
Some suggestions for the tests

christian-byrne committed:
[test] extract createMockWidget to shared test utility - addresses @DrJKL's code replication concern

Creates testUtils.ts with shared createMockWidget and createMockFile functions
to reduce duplication across widget component tests. This ensures consistency
and maintainability of test setup code.

christian-byrne committed:
[test] replace type assertions with type narrowing - addresses @DrJKL's type safety suggestion

Replaces unsafe `as HTMLInputElement` casts with proper instanceof checks
and error throwing. Also refactors File Type Detection tests to use it.for
instead of conditionals to eliminate anti-pattern.

christian-byrne committed:
[feat] use destructuring with default value for readonly prop - addresses @DrJKL's Vue best practice suggestion

Replace manual fallback expressions like `readonly || false` with modern Vue 3
destructuring pattern: `const { readonly = false } = defineProps()`.
This is cleaner than withDefaults() and follows current Vue best practices.

christian-byrne committed:
[test] improve test utilities usage - addresses @DrJKL's additional suggestions

- Replace findComponent with getComponent for better error handling
- Use optional chaining (?.()) instead of conditional checks for cleaner syntax
- Remove unnecessary existence checks since getComponent throws on failure

christian-byrne commented:
Use destructuring for props https://github.com/Comfy-Org/ComfyUI_frontend/pull/5528/commits/7e7748c31a9eb0bfa1811162df236ee300fae601
